hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/ContainerEndpoint.java

```diff
@@ -481,4 +489,83 @@ private List<ContainerBlockMetadata> getBlocks(
     return blockIds;
   }
 
+  @GET
+  @Path("/mismatch")
+  public Response getContainerMisMatchInsights() {
+    List<ContainerDiscrepancyInfo> containerDiscrepancyInfoList =
+        new ArrayList<>();
+    try {
+      Map<Long, ContainerMetadata> omContainers =
+          reconContainerMetadataManager.getContainers(-1, -1);
+      List<Long> scmAllContainers = containerManager.getContainers().stream()
+          .map(containerInfo -> containerInfo.getContainerID()).collect(
+              Collectors.toList());
+      List<Long> scmNonDeletedContainers =
+          containerManager.getContainers().stream()
+              .filter(containerInfo -> !(containerInfo.getState() ==
+                  HddsProtos.LifeCycleState.DELETED))
+              .map(containerInfo -> containerInfo.getContainerID()).collect(
+                  Collectors.toList());
+
+      // Filter list of container Ids which are present in OM but not in SCM.
+      List<Map.Entry<Long, ContainerMetadata>> notSCMContainers =
+          omContainers.entrySet().stream().filter(containerMetadataEntry ->
+                  !(scmAllContainers.contains(containerMetadataEntry.getKey())))
```

Comment:
This still looks incorrect. If a container has got Deleted in SCM and still present in OM, it will never get flagged as scmAllContainers has Deleted SCM containers as well.  so no discrepancy will ever get reported.
you need to have  scmNonDeletedContainers here as well instead of scmAllContainers. What we are trying to figure out is container is still referred in OM but in SCM that container has got deleted.
@devmadhuu @sumitagrawl

Reply:
> This still looks incorrect. If a container has got Deleted in SCM and still present in OM, it will never get flagged as scmAllContainers has Deleted SCM containers as well. so no discrepancy will ever get reported. you need to have scmNonDeletedContainers here as well instead of scmAllContainers. What we are trying to figure out is container is still referred in OM but in SCM that container has got deleted. @devmadhuu @sumitagrawl

IMO, we want to identify mismatch of container referred in OM but present in SCM, now why the container is in deleted state in SCM, that is different issue. But any container which cannot be queried for metadata by OM to SCM is a data loss situation. @sumitagrawl - I think your 1st point is what we discussed. Let us know what discussion happened with @errose28

Reply:
@devmadhuu We should not count deleted SCM container for comparing to OM also, deleted container is just a bookmark, to avoid re-created same container at DN due to issue with sync at DN or Old DN.

This needs to be removed in comparison here tool.